packages/babel-traverse/src/path/family.js

```diff
@@ -18,6 +18,34 @@ function addCompletionRecords(path, paths) {
   return paths;
 }
 
+function completionRecordForSwitch(cases, paths) {
+  for (const switchCase of cases) {
+    let consequent = switchCase.get("consequent");
+    if (consequent.length > 0 && consequent[0].isBlockStatement()) {
+      consequent = consequent[0].get("body");
+    }
+    const isDefaultCase = switchCase.isSwitchCase({ test: null });
+    const breakStatement = consequent.find(statement =>
+      statement.isBreakStatement(),
+    );
+
+    if (breakStatement) {
+      // get the previous statement
+      const previousKey = breakStatement.key - 1;
+      if (previousKey >= 0 && consequent[previousKey].isExpressionStatement()) {
+        paths = addCompletionRecords(consequent[previousKey], paths);
+        breakStatement.remove();
+      } else {
+        breakStatement.replaceWith(switchCase.scope.buildUndefinedNode());
+        paths = addCompletionRecords(breakStatement, paths);
+      }
+    } else if (consequent.length > 0 && isDefaultCase) {
```

Comment:
`default` cases have nothing special, they should be handled exactly like `case`:

This should evaluate to `1`:
```js
const a = do {
  switch (2) {
    default: 0;
    case 1: 1; break;
  }
};
```

This should evaluate to `1`:
```js
const a = do {
  switch (1) {
    case 1: 1;
  }
};
```

The check should probably be more like `if (isLastStatement) {`

Reply:
Also this should evaluate to 1:
```js
const a = do {
  switch (1) {
    case 1: 1;
    case 2:
  }
};
```

Reply:
So can I say that the lastStatement is an expression statement , then no fall through, else will fall through say:

```
let a = 1
const b = do {
switch (1) {
   case 1: a = 2;
   case 2: a * 10;
}
```

So b = ?

Reply:
10, because the last statement is `a*10`

Reply:
ok. interesting. is this documented somewhere within https://github.com/tc39/proposal-do-expressions? or where do I find the spec for this?

Reply:
1. Switch expressions evaluate to the result of their block: https://tc39.es/ecma262/#sec-switch-statement-runtime-semantics-evaluation steps 7 and 9
2. It delegates to https://tc39.es/ecma262/#sec-runtime-semantics-caseblockevaluation, in the `CaseBlock: { CaseClauses }` section.
  i. At the first iteration of step 4 of that algorithm, `V` is set to `2` (the result of `a = 2`).
  ii. Since `case 1` isn't followed by a break/return/throw statement, `R` is a NormalCompletion and not an AbruptCompletion: for this reason the loop continues running to the following case.
  iii. At the second iteration, `V` is set to `20` (the result of `a * 10`).
  iv. The loop ends, because there are no more `case`s, and it returns `20`

Reply:
:man_facepalming: :man_facepalming: :man_facepalming: :man_facepalming: 
I just noticed I wrote 10 instead of 20

Reply:
I realised reading the ECMA262 Spec is giving me more confusion 🤦‍♂ 🤦‍♂ 🤦‍♂, but I guess I roughly know what you trying to point out. And I think we should document more in the do-expression proposal docs

Reply:
Btw, you can already "test" do expression in any browser.
```js
do { ANYTHING }
```
is
```js
eval(`{
  ANYTHING
}`);